pkg/build/nodeimage/helpers.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nodeimage
+
+import (
+	"fmt"
+	"path"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/util/version"
+
+	"sigs.k8s.io/kind/pkg/exec"
+)
+
+// createFile creates the file at filePath in the container,
+// ensuring the directory exists and writing contents to the file
+func createFile(containerCmder exec.Cmder, filePath, contents string) error {
+	// NOTE: the paths inside the container should use the path package
+	// and not filepath (!), we want posixy paths in the linux container, NOT
+	// whatever path format the host uses. For paths on the host we use filepath
+	if err := containerCmder.Command("mkdir", "-p", path.Dir(filePath)).Run(); err != nil {
+		return err
+	}
+
+	return containerCmder.Command(
+		"cp", "/dev/stdin", filePath,
+	).SetStdin(
+		strings.NewReader(contents),
+	).Run()
+}
+
+func repositoryCorrectorForVersion(kubeVersion *version.Version, arch string) func(string) string {
+	archSuffix := "-" + arch
+
+	// For kubernetes v1.15+ (actually 1.16 alpha versions) we may need to
+	// drop the arch suffix from images to get the expected image
+	// for < v1.12 we need to do the opposite.
+	// We can accomplish this by just handling < 1.12 & >= 1.12 as we won't
+	// touch images that match the expectation in either case ...
+	if kubeVersion.LessThan(version.MustParseSemantic("v1.12.0")) {
+		return func(repository string) string {
+			if !strings.HasSuffix(repository, archSuffix) {
+				fixed := repository + archSuffix
+				fmt.Println("fixed: " + repository + " -> " + fixed)
+				repository = fixed
+			}
+			return repository
+		}
+	}
+
+	return func(repository string) string {
+		if strings.HasSuffix(repository, archSuffix) {
+			fixed := strings.TrimSuffix(repository, archSuffix)
+			fmt.Println("fixed: " + repository + " -> " + fixed)
+			repository = fixed
+		}
+		return repository
+	}
+}
+
+func findSandboxImage(images []string) string {
+	for _, image := range images {
+		// yep this seems legit
+		// https://github.com/kubernetes-sigs/kind/issues/1471#issuecomment-617579803
+		if strings.Contains(image, "pause") {
```

Comment:
It seems you call this function with the output of 
`ctr", "--namespace=k8s.io", "images", "list", "-q"`, that gives me

```
root@kind-control-plane:/# ctr --namespace=k8s.io images list -q
docker.io/kindest/kindnetd:0.5.4
docker.io/rancher/local-path-provisioner:v0.0.12
k8s.gcr.io/coredns:1.6.7
k8s.gcr.io/debian-base:v2.0.0
k8s.gcr.io/etcd:3.4.7-0
k8s.gcr.io/kube-apiserver:v1.19.0-alpha.1.677_e95106106b1e24-dirty
k8s.gcr.io/kube-controller-manager:v1.19.0-alpha.1.677_e95106106b1e24-dirty
k8s.gcr.io/kube-proxy:v1.19.0-alpha.1.677_e95106106b1e24-dirty
k8s.gcr.io/kube-scheduler:v1.19.0-alpha.1.677_e95106106b1e24-dirty
k8s.gcr.io/pause:3.1
k8s.gcr.io/pause:3.2
```

it will pick the 3.1, and the issue that you are referring is using
```
kubeadm config images list | grep pause
```

that gives me
```
k8s.gcr.io/pause:3.2
```

Reply:
Hi, are you running this with a node image built from this PR? 🙃

Reply:
If you use a node image built before this PR against kubernetes that prefers 3.2 in kubeadm then the node image has 3.2 but a running cluster pulls 3.1

With this PR container gets configured to 3.2 or 3.1 based on kubeadm at build time.

There's only one.

Reply:
ic, but then why is not better directly the output of `kubeadm config images list | grep pause
` ? is less prone to

Reply:
I don't see why it's less prone to future errors and there's no error in what's here.

Reply:
Remember, we are not running this on a live cluster. This is listing the images loaded into the node after image loading is done

Reply:
my rationale is that  `ctr", "--namespace=k8s.io", "images", "list", "-q"` list whatever containerd has imported, and `kubeadm config images list` list whatever `kubeadm` has specified :smile: 

It is not a big deal, but who knows if there will be an image preloaded ... maybe just too paranoid

Reply:
at some point we may not follow kubeadm (e.g. if we start skipping preflight) but we will always need the image to be preloaded

Reply:
got it 👍